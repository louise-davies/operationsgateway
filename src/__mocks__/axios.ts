// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

const requests = {
  get: jest.fn((path) => {
    if (path === '/operationsgateway-settings.json') {
      return Promise.resolve({
        data: {
          apiUrl: 'api',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
};

export default requests;
