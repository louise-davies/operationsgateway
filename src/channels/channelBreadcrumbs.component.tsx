import { NavigateNext } from '@mui/icons-material';
import { Breadcrumbs, Link, Typography } from '@mui/material';
import React from 'react';

type ChannelBreadcrumbsProps = {
  currNode: string;
  setCurrNode: (newNode: string) => void;
};

const ChannelBreadcrumbs = (props: ChannelBreadcrumbsProps) => {
  const { currNode, setCurrNode } = props;
  // special handling for '/' otherwise we get ['','']
  const nodePath = currNode === '/' ? [''] : currNode.split('/');

  return (
    <Breadcrumbs
      separator={<NavigateNext fontSize="small" />}
      aria-label="breadcrumb"
    >
      {nodePath.map((value, currIndex) => {
        const label = value.length ? value : 'All Channels';
        // final node, render as non-"link"
        if (currIndex === nodePath.length - 1) {
          return (
            <Typography key={label} color="text.primary">
              {label}
            </Typography>
          );
        } else {
          return (
            <Link
              key={label}
              href="#"
              onClick={(ev) => {
                // prevent default to stop the href="#" adding a hash to the URL
                // and potentially jumping the page
                ev.preventDefault();
                setCurrNode(nodePath.slice(0, currIndex + 1).join('/'));
              }}
            >
              {label}
            </Link>
          );
        }
      })}
    </Breadcrumbs>
  );
};

export default ChannelBreadcrumbs;
