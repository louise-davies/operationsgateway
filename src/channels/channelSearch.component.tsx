import { Search } from '@mui/icons-material';
import {
  Autocomplete,
  type FilterOptionsState,
  InputAdornment,
  TextField,
} from '@mui/material';
import React from 'react';
import { FullChannelMetadata } from '../app.types';

type ChannelSearchProps = {
  currPathAndChannel: string;
  channels: FullChannelMetadata[];
  onSearchChange: (channel: FullChannelMetadata) => void;
};

const filterOptions = (
  options: FullChannelMetadata[],
  { inputValue }: FilterOptionsState<FullChannelMetadata>
) => {
  const searchString = inputValue.toLowerCase();

  return options.filter((option) => {
    const name = option?.name?.toLowerCase();
    const systemName = option.systemName.toLowerCase();
    // need to search both friendly and system names for a match
    return name?.includes(searchString) || systemName.includes(searchString);
  });
};

const ChannelSearch = (props: ChannelSearchProps) => {
  const { onSearchChange, channels, currPathAndChannel } = props;

  const [value, setValue] = React.useState<FullChannelMetadata | null>(null);
  const [inputValue, setInputValue] = React.useState('');

  React.useEffect(() => {
    // clear the search box if the user clicks another channel / navigates via breadcrumbs
    if (value && currPathAndChannel !== `${value.path}/${value.systemName}`) {
      setValue(null);
      setInputValue('');
    }
  }, [currPathAndChannel, value]);

  return (
    <Autocomplete
      fullWidth
      filterOptions={filterOptions}
      value={value}
      inputValue={inputValue}
      onInputChange={(event, newInputValue) => setInputValue(newInputValue)}
      size="small"
      options={channels}
      getOptionLabel={(option) =>
        option.name
          ? `${option.name} (${option.systemName})`
          : option.systemName
      }
      blurOnSelect
      onChange={(event: unknown, newValue: FullChannelMetadata | null) => {
        if (newValue) onSearchChange(newValue);
        setValue(newValue);
      }}
      renderInput={(params) => (
        <TextField
          {...params}
          label="Search data channels"
          InputProps={{
            ...params.InputProps,
            startAdornment: (
              <InputAdornment position="start">
                <Search />
              </InputAdornment>
            ),
          }}
        />
      )}
    />
  );
};

export default ChannelSearch;
