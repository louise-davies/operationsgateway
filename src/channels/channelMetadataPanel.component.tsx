import AddIcon from '@mui/icons-material/Add';
import RemoveIcon from '@mui/icons-material/Remove';
import {
  Button,
  Divider,
  Stack,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from '@mui/material';
import React from 'react';
import { useChannelSummary } from '../api/channels';
import {
  FullChannelMetadata,
  isChannelMetadataScalar,
  isChannelMetadataWaveform,
} from '../app.types';
import {
  TraceOrImageThumbnail,
  renderTimestamp,
} from '../table/cellRenderers/cellContentRenderers';

type ChannelMetadataPanelProps = {
  displayedChannel: FullChannelMetadata | undefined;
  isChannelSelected: boolean;
  onSelectChannel: (channel: string) => void;
  onDeselectChannel: (channel: string) => void;
};

const Heading = (props: React.ComponentProps<typeof Typography>) => {
  const { children, ref, ...restProps } = props;
  return (
    <Typography
      variant="body1"
      component="h3"
      gutterBottom
      sx={{ fontWeight: 'bold' }}
      {...restProps}
    >
      {children}
    </Typography>
  );
};
const Body = (props: React.ComponentProps<typeof Typography>) => {
  const { children, ...restProps } = props;
  return (
    <Typography variant="body2" gutterBottom {...restProps}>
      {props.children}
    </Typography>
  );
};

const ChannelMetadataPanel = (props: ChannelMetadataPanelProps) => {
  const {
    displayedChannel,
    isChannelSelected,
    onSelectChannel,
    onDeselectChannel,
  } = props;

  const { data: channelSummary } = useChannelSummary(
    displayedChannel?.systemName
  );

  function addCurrentChannel() {
    if (displayedChannel?.systemName) {
      onSelectChannel(displayedChannel.systemName);
    }
  }

  function removeCurrentChannel() {
    if (displayedChannel?.systemName) {
      onDeselectChannel(displayedChannel.systemName);
    }
  }

  if (displayedChannel) {
    return (
      <>
        <Stack
          direction="row"
          justifyContent="space-between"
          alignItems="center"
        >
          <Heading gutterBottom={false}>
            {displayedChannel?.name ?? displayedChannel.systemName}
          </Heading>
          {isChannelSelected ? (
            <Button
              variant="outlined"
              startIcon={<RemoveIcon />}
              size="small"
              onClick={removeCurrentChannel}
            >
              Remove this channel
            </Button>
          ) : (
            <Button
              variant="contained"
              startIcon={<AddIcon />}
              size="small"
              onClick={addCurrentChannel}
            >
              Add this channel
            </Button>
          )}
        </Stack>
        {displayedChannel?.name && (
          <Body>System name: {displayedChannel.systemName}</Body>
        )}
        {displayedChannel?.description && (
          <Body>{displayedChannel.description}</Body>
        )}
        <Body>Channel type: {displayedChannel.type}</Body>
        {isChannelMetadataScalar(displayedChannel) &&
          displayedChannel.units && (
            <Body>Units: {displayedChannel.units}</Body>
          )}
        {isChannelMetadataWaveform(displayedChannel) && (
          <>
            <Body gutterBottom={false}>
              X Units: {displayedChannel.x_units}
            </Body>
            <Body>Y Units: {displayedChannel.y_units}</Body>
          </>
        )}
        {displayedChannel?.historical && (
          <Body fontWeight="bold">This channel is historical</Body>
        )}
        {channelSummary && (
          <>
            <Divider />
            <Typography
              variant="body2"
              component="h4"
              gutterBottom
              sx={{ fontWeight: 'bold', paddingTop: 1 }}
            >
              Data Summary
            </Typography>
            <Body gutterBottom={false}>
              First data date: {renderTimestamp(channelSummary.first_date)}
            </Body>
            <Body>
              Most recent data date:{' '}
              {renderTimestamp(channelSummary.most_recent_date)}
            </Body>
            <Typography
              variant="body2"
              component="h5"
              gutterBottom
              sx={{ fontWeight: 'bold', paddingTop: 1 }}
            >
              Recent Data
            </Typography>
            <TableContainer>
              <Table aria-label="recent data" size="small">
                <TableHead>
                  <TableRow>
                    <TableCell>Time</TableCell>
                    <TableCell>Data</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {channelSummary.recent_sample.map((sample) => {
                    const [timestamp, data] = Object.entries(sample)[0];
                    const formattedTimestamp = renderTimestamp(timestamp);
                    return (
                      <TableRow
                        key={formattedTimestamp}
                        sx={{
                          '&:last-child td, &:last-child th': { border: 0 },
                        }}
                      >
                        <TableCell component="th" scope="row">
                          {formattedTimestamp}
                        </TableCell>
                        <TableCell>
                          {isChannelMetadataScalar(displayedChannel) ? (
                            data
                          ) : (
                            <TraceOrImageThumbnail
                              base64Data={data as string}
                              alt={`${
                                displayedChannel?.name ??
                                displayedChannel.systemName
                              } ${
                                displayedChannel.type
                              } at ${formattedTimestamp}`}
                            />
                          )}
                        </TableCell>
                      </TableRow>
                    );
                  })}
                </TableBody>
              </Table>
            </TableContainer>
          </>
        )}
      </>
    );
  } else {
    return (
      <>
        <Heading>Data Channels Help</Heading>
        <Body>
          There are two options to locate a channel. You can either search or
          use the navigation on the left.
        </Body>
        <Body>
          Channel metadata can also be viewed. This can be displayed by clicking
          a channel name
        </Body>
      </>
    );
  }
};

export default ChannelMetadataPanel;
