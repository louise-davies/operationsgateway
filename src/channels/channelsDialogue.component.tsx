import React from 'react';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Divider,
  Grid,
  InputAdornment,
  TextField,
} from '@mui/material';
import { Search } from '@mui/icons-material';
import { useChannels } from '../api/channels';
import { FullChannelMetadata } from '../app.types';
import ChannelTree from './channelTree.component';
import ChannelBreadcrumbs from './channelBreadcrumbs.component';
import { useAppDispatch, useAppSelector } from '../state/hooks';
import {
  selectSelectedIdsIgnoreOrder,
  updateSelectedColumns,
} from '../state/slices/tableSlice';
import { createSelector } from '@reduxjs/toolkit';

interface ChannelsDialogueProps {
  open: boolean;
  onClose: () => void;
}

export interface TreeNode {
  name: string;
  checked: boolean;
  children: Record<
    string,
    TreeNode | (FullChannelMetadata & { checked: boolean })
  >;
}

/**
 * @returns A selector for a tree representing the channel hierarchy,
 * which is used by components in the channels folder
 * @params state - redux state - isn't actually used! (means we get nice memoisation from reselect)
 * @params availableChannels - array of all the channels the user can select
 * @params selectedIds - array of all the channels currently selected
 */
export const selectChannelTree = createSelector(
  (
    state: unknown,
    availableChannels: FullChannelMetadata[],
    selectedIds: string[]
  ) => availableChannels,
  (
    state: unknown,
    availableChannels: FullChannelMetadata[],
    selectedIds: string[]
  ) => selectedIds,
  (availableChannels, selectedIds) => {
    const tree: TreeNode = { name: '/', children: {}, checked: false };
    availableChannels.forEach((channel) => {
      const { path } = channel;
      // split the path and remove any empty strings
      const splitPath = path.split('/').filter((el) => el);
      splitPath.reduce((prev, curr, currIndex) => {
        // init treenode if not already initialised
        if (!prev.children?.[curr]) {
          prev.children[curr] = {
            name: curr,
            checked: false,
            children: {},
          };
        }
        const node = prev.children[curr] as TreeNode;
        // last item, so therefore is the channel
        if (currIndex === splitPath.length - 1) {
          node.children[channel.systemName] = {
            ...channel,
            checked: selectedIds.includes(channel.systemName),
          };
        }
        // return the child node, so we drill deeper into the tree
        return node;
      }, tree);
    });

    return tree;
  }
);

const ChannelsDialogue = (props: ChannelsDialogueProps) => {
  const { open, onClose } = props;

  const { data: channels } = useChannels();

  const appliedSelectedIds = useAppSelector((state) =>
    selectSelectedIdsIgnoreOrder(state)
  );

  const [selectedIds, setSelectedIds] = React.useState(appliedSelectedIds);

  React.useEffect(() => {
    setSelectedIds(appliedSelectedIds);
  }, [appliedSelectedIds]);

  const channelTree = useAppSelector((state) =>
    selectChannelTree(state, channels ?? [], selectedIds)
  );

  const dispatch = useAppDispatch();

  const onChannelSelect = React.useCallback((channel: string): void => {
    setSelectedIds((selectedIds) => [...selectedIds, channel]);
  }, []);

  const onChannelDeselect = React.useCallback((channel: string): void => {
    setSelectedIds((selectedIds) =>
      selectedIds.filter((colId) => colId !== channel)
    );
  }, []);

  const handleChannelChecked = React.useCallback(
    (channel: string, checked: boolean) => {
      checked ? onChannelDeselect(channel) : onChannelSelect(channel);
    },
    [onChannelDeselect, onChannelSelect]
  );

  const [currNode, setCurrNode] = React.useState('/');

  return (
    <Dialog open={open} onClose={onClose} maxWidth="lg" fullWidth>
      <Grid container columnSpacing={2} alignItems="center">
        <Grid item xs>
          <DialogTitle>Data Channels</DialogTitle>
        </Grid>
        <Grid item xs>
          <TextField
            size="small"
            label="Search data channels"
            fullWidth
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  <Search />
                </InputAdornment>
              ),
            }}
          />
        </Grid>
      </Grid>
      <DialogContent
        sx={{
          border: 'thin lightgrey',
          borderStyle: 'solid none',
          paddingTop: '8px !important',
          paddingBottom: '8px',
          overflowY: 'unset',
        }}
      >
        <ChannelBreadcrumbs currNode={currNode} setCurrNode={setCurrNode} />
      </DialogContent>
      <DialogContent>
        <Grid container columnSpacing={2}>
          <Grid item xs>
            <ChannelTree
              currNode={currNode}
              tree={channelTree ?? { name: '/', children: {}, checked: false }}
              setCurrNode={setCurrNode}
              handleChannelChecked={handleChannelChecked}
            />
          </Grid>
          <Divider orientation="vertical" flexItem />
          <Grid item xs>
            Help / info
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Close</Button>
        <Button
          onClick={() => {
            dispatch(updateSelectedColumns(selectedIds));
            onClose();
          }}
        >
          Add Channels
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default ChannelsDialogue;
