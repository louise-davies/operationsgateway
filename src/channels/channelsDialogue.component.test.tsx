import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import ChannelsDialogue, {
  selectChannelTree,
  TreeNode,
} from './channelsDialogue.component';
import {
  getInitialState,
  renderComponentWithProviders,
  testChannels,
} from '../setupTests';
import { PreloadedState } from '@reduxjs/toolkit';
import axios from 'axios';
import { QueryClient } from '@tanstack/react-query';
import { RootState } from '../state/store';
import { staticChannels } from '../api/channels';

describe('selectChannelTree', () => {
  it('transforms channel list with selection info into TreeNode', () => {
    const channels = testChannels;
    channels[5].path = '/test_1';
    channels[6].path = '/test_3/test_3_sub';
    const selectedIds = ['timestamp', 'shotnum', 'test_1'];
    const channelTree = selectChannelTree({}, channels ?? [], selectedIds);

    const expectedTree: TreeNode = {
      name: '/',
      checked: false,
      children: {
        system: {
          name: 'system',
          checked: false,
          children: Object.fromEntries(
            Object.entries(staticChannels).map(([channelName, channel]) => [
              channelName,
              {
                ...channel,
                checked: selectedIds.find((channel) => channel === channelName)
                  ? true
                  : false,
              },
            ])
          ),
        },
        test_1: {
          name: 'test_1',
          checked: false,
          children: {
            [testChannels[4].systemName]: {
              ...testChannels[4],
              checked: true,
            },
            [testChannels[5].systemName]: {
              ...testChannels[5],
              checked: false,
            },
          },
        },
        test_3: {
          name: 'test_3',
          checked: false,
          children: {
            test_3_sub: {
              name: 'test_3_sub',
              checked: false,
              children: {
                [testChannels[6].systemName]: {
                  ...testChannels[6],
                  checked: false,
                },
              },
            },
          },
        },
      },
    };

    expect(channelTree).toStrictEqual(expectedTree);
  });
});

describe('Channels Dialogue', () => {
  let props: React.ComponentProps<typeof ChannelsDialogue>;
  let user;

  const createView = (
    initialState?: PreloadedState<RootState>,
    queryClient?: QueryClient
  ) => {
    return renderComponentWithProviders(<ChannelsDialogue {...props} />, {
      preloadedState: initialState,
      queryClient,
    });
  };

  beforeEach(() => {
    user = userEvent.setup();
    props = {
      open: true,
      onClose: jest.fn(),
    };

    (axios.get as jest.Mock).mockResolvedValue({
      data: { channels: testChannels.slice(4) },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders channels dialogue when dialogue is open', async () => {
    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });
    // wait for the channels to load before taking the screenshot
    await screen.findByText('system');
    expect(baseElement).toMatchSnapshot();
  });

  it("doesn't render channels dialogue when dialogue is closed", async () => {
    props.open = false;

    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });

  it('calls onClose when close button is clicked', async () => {
    createView();

    await user.click(screen.getByText('Close'));

    expect(props.onClose).toHaveBeenCalled();
  });

  it('dispatches updateSelectedColumns and onClose when apply button is clicked', async () => {
    const state = {
      ...getInitialState(),
      table: {
        ...getInitialState().table,
        selectedColumnIds: ['timestamp', 'activeArea', 'test_1'],
      },
    };

    const { store } = createView(state);

    await user.click(await screen.findByText('system'));

    expect(screen.getByRole('checkbox', { name: 'Active Area' })).toBeChecked();
    expect(
      screen.getByRole('checkbox', { name: 'Shot Number' })
    ).not.toBeChecked();

    await user.click(screen.getByRole('checkbox', { name: 'Shot Number' }));
    await user.click(screen.getByRole('checkbox', { name: 'Active Area' }));

    await user.click(screen.getByText('Add Channels'));

    expect(store.getState().table.selectedColumnIds).toStrictEqual([
      'timestamp',
      'test_1',
      'shotnum',
    ]);
    expect(props.onClose).toHaveBeenCalled();
  });
});
