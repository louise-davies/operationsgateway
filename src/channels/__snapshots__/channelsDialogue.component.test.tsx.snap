// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channels Dialogue doesn't render channels dialogue when dialogue is closed 1`] = `
<body
  style=""
>
  <div />
</body>
`;

exports[`Channels Dialogue renders channels dialogue when dialogue is open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthLg MuiDialog-paperFullWidth css-cyxlny-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-sergpw-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
              id=":r0:"
            >
              Data Channels
            </h2>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-n4yv6w-MuiGrid-root"
          >
            <div
              class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-12szb0g-MuiAutocomplete-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r1:"
                  id=":r1:-label"
                >
                  Search data channels
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-ittuaa-MuiInputAdornment-root"
                  >
                    <span
                      class="notranslate"
                    >
                      â€‹
                    </span>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="SearchIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                      />
                    </svg>
                  </div>
                  <input
                    aria-autocomplete="list"
                    aria-expanded="false"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-12yjm75-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    role="combobox"
                    spellcheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Open"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      tabindex="-1"
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        Search data channels
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogContent-root css-1iby97h-MuiDialogContent-root"
        >
          <nav
            aria-label="breadcrumb"
            class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
          >
            <ol
              class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
            >
              <li
                class="MuiBreadcrumbs-li"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
                >
                  All Channels
                </p>
              </li>
            </ol>
          </nav>
        </div>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container css-o1xwpy-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
            >
              <ul
                class="MuiList-root MuiList-dense css-1xbd29p-MuiList-root"
              >
                <li
                  class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
                >
                  <div
                    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                    >
                      <span
                        aria-disabled="true"
                        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                        tabindex="-1"
                      >
                        <input
                          aria-labelledby="checkbox-list-label-system"
                          class="PrivateSwitchBase-input css-1m9pwf3"
                          data-indeterminate="false"
                          disabled=""
                          type="checkbox"
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                          data-testid="CheckBoxOutlineBlankIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                    </div>
                    <div
                      class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
                      id="checkbox-list-label-system"
                    >
                      <span
                        class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                      >
                        system
                      </span>
                    </div>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </div>
                </li>
                <li
                  class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
                >
                  <div
                    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                    >
                      <span
                        aria-disabled="true"
                        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                        tabindex="-1"
                      >
                        <input
                          aria-labelledby="checkbox-list-label-Channels"
                          class="PrivateSwitchBase-input css-1m9pwf3"
                          data-indeterminate="false"
                          disabled=""
                          type="checkbox"
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                          data-testid="CheckBoxOutlineBlankIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                    </div>
                    <div
                      class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
                      id="checkbox-list-label-Channels"
                    >
                      <span
                        class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                      >
                        Channels
                      </span>
                    </div>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </div>
                </li>
              </ul>
            </div>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-hbvkri-MuiGrid-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-um3j9i-MuiTypography-root"
              >
                Data Channels Help
              </h3>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ejgmvm-MuiTypography-root"
              >
                There are two options to locate a channel. You can either search or use the navigation on the left.
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ejgmvm-MuiTypography-root"
              >
                Channel metadata can also be viewed. This can be displayed by clicking a channel name
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Add Channels
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
